### PR TITLE
feat: add Docker Compose deployment, refactor config paths and mock modes to CLI

### PR DESCRIPTION
**Pull request by hwei0**
# Description

This PR adds a full Docker Compose deployment setup for TURBO, refactors how config paths (ZMQ sockets, log directories) are resolved at runtime, and moves mock mode toggles from YAML config fields to CLI flags.

### Pre-built Docker image workflow (new)
- Adds a root-level `compose.yaml` and `.env.example` that pull pre-built images from [DockerHub](https://hub.docker.com/u/hbwei), eliminating the need for local builds, SSL key generation, or Rust toolchain installation
- The Quick Start now defaults to `docker compose pull` + `docker compose up` (no `--build`), reducing setup to: clone, download data, configure `.env`, and run
- The existing `docker/` directory (Dockerfiles, build-oriented `compose.yaml`, `docker/.env.example`) is preserved as the "build from source" workflow for development/customization
- `docker/README.Docker.md` rewritten to focus on building from source as the secondary path, with shared config reference and troubleshooting kept in place

### Docker Compose deployment
- Adds 4 multi-stage Dockerfiles (Python base/binary, Rust base, QUIC binary) and a `compose.yaml` orchestrating 7 services across `client` and `server` profiles
- Services start in dependency order via health checks: `client_python_main` writes `/health/client_main_ready` (only after all Client processes have bound their `quic_rcv_zmq_socket`), then `client_python_monitor` writes `/health/monitor_ready`, then `quic_client` starts. This is coordinated via a `multiprocessing.Manager().Queue()` for cross-process ZMQ bind readiness
- All services use `ipc: host` for ZeroMQ IPC and POSIX shared memory across containers, `init: true` (tini) for signal forwarding, and a custom bridge network (`10.64.89.0/24`) for QUIC UDP
- Both Python orchestrators now handle SIGTERM in addition to SIGINT for Docker graceful shutdown
- Docker-specific YAML configs (`docker/config/`) mirror the manual configs with container-internal paths (`/app/...`)
- `.env.example` provides all configurable variables (host UID/GID, model/eval data paths, mock mode toggles, networking)
- Docker Compose Watch support for hot-reload during development

### GPU support made optional via compose override
- Removed hardcoded NVIDIA GPU reservations from both `client_python_main` (never needed — client uses PyTorch for CPU tensor ops only, no `torch.cuda` usage) and `server_python_main`
- GPU reservations for the server are now in separate `compose.gpu.yaml` override files (root and `docker/`), included only on GPU hosts via `-f compose.gpu.yaml`
- Non-GPU hosts can run the server with mock inference without the NVIDIA Container Toolkit: `docker compose --profile server up`
- GPU hosts include the override: `docker compose -f compose.yaml -f compose.gpu.yaml --profile server up`
- Fixes `Error response from daemon: could not select device driver "nvidia" with capabilities: [[gpu]]` on non-GPU hosts

### Documentation improvements
- Updated `README.md` to recommend pre-built Docker images as the default, with detailed prerequisites including:
  - Docker 28 incompatibility warning (`nf_tables`/iptables issue; Docker 27 recommended)
  - Disk space requirements (client ~50 GB, server ~13 GB)
  - Clarified that NVIDIA GPU drivers + Container Toolkit are **server-only** (not needed for client or mock inference)
  - USB webcam prerequisites (**client-only**, with mock camera alternative)
  - Docker Compose V2 plugin installation instructions
  - "No GPU?" guidance for running with `MOCK_INFERENCE=true`
- Added per-step annotations for multi-machine deployments ("both client and server", "server only", "client only") across all setup steps
- `.env` table now shows quickstart default values (`~/av-models`, `~/full-eval`, `~/experiment2-out`) instead of "(must set)"
- Clarified mock mode data requirements: `full-eval` evaluation data is always required on the client (even in mock camera mode — the bandwidth allocator needs it for utility curve computation); `av-models` model checkpoints are not needed in mock inference mode
- Added "Data Required" column to the mock mode combination table
- Manual (non-Docker) setup collapsed into a `<details>` block with deduplicated prerequisites
- `docker/README.Docker.md` streamlined — duplicated setup steps replaced with cross-references to main README; removed erroneous Rust host prerequisite (Rust compiles inside the Docker build stage)

### Config path refactoring (Python + Rust + YAML)
- **ZMQ socket paths**: Changed from hardcoded full `ipc:///absolute/path/socket-name` to bare names (e.g., `service1-camera-socket`) in all YAML configs. The Python orchestrators (`client_main.py`, `server_main.py`) and web dashboard (`web_config.py`) resolve them at runtime to `ipc://<zmq_dir>/<name>`. The Rust QUIC binaries already used this pattern and were updated to match the renamed config key (`zmq_pathdir` → `zmq_dir`)
- **Log output directories**: Removed per-component `client_savedir`, `camera_savedir`, `ping_savedir`, `server_log_savedir`, `quic_client_log_path`, and `quic_server_log_path` from YAML. Each orchestrator now auto-creates a timestamped run directory (e.g., `experiment_output_dir/client_main_2024-01-15_10-30-00/client/`) and injects the save path into component configs at startup. The Rust QUIC binaries now also create timestamped subdirectories using `chrono`
- **Server IP for ping**: Removed `DST_IP` YAML field. Server IP is now extracted from the new required `-s/--server_address` CLI argument on `client_main.py`
- New top-level config keys: `experiment_output_dir`, `zmq_dir`, `client_subdir`/`server_subdir`, `quic_client_log_subdir`/`quic_server_log_subdir`

### Mock mode refactoring
- Mock camera (`client_main.py --mock-camera`) and mock inference (`server_main.py --mock-inference`) are now toggled via CLI flags instead of setting `null` vs a path in YAML. The YAML always specifies the mock data paths; the CLI flag controls whether they're used
- In Docker, mock modes are toggled via `MOCK_CAMERA`/`MOCK_INFERENCE` environment variables in `.env`
- Added `mock_model_latency_csv_path` to `ModelServerConfig` — mock inference now simulates realistic per-model latency from `experiment_model_info.csv` instead of a hardcoded 50ms sleep
- Renamed mock output file from `.npz` to `.npy` and switched from `np.fromfile()` to `np.load()`

### Other changes
- **Cargo.lock now tracked in version control** — removed `**/Cargo.lock` from `.gitignore` and committed the lockfile to ensure reproducible Rust builds across environments
- Added panic hooks with backtrace capture in both Rust QUIC binaries
- Error callbacks in Python orchestrators now log full stack traces
- `start_web_dashboard.py`: Added `-c` short flag for `--config`
- `web_frontend.py`: Added `allow_unsafe_werkzeug=True` for running inside Docker containers
- Dockerfile UID/GID defaults normalized to `1000` across all images; `groupadd` moved from base to binary stage
- Updated `CONFIGURATION.md`, `IPC.md`, and `LOGGING.md` to reflect all the above changes

Fixes # (issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Tested successful connectivity using mock gpus and mock cameras on google cloud. Also tested successful connectivity using mock modes from laptop to google cloud. 

# Checklist:

- [x] My commit title follows [conventional commit guidelines](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I have made corresponding changes to the documentation
- [x] I have tested my changes, confirming that my feature works

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -91,56 +91,46 @@ Running on a GPU-equipped cloud instance (e.g., H100):
...
+The recommended way to run TURBO is with Docker. The Docker setup automatically orchestrates all processes — 2 on the server (QUIC server + model servers) and 3 on the client (client orchestrator + web dashboard + QUIC client) — handling startup ordering, ZMQ socket management, and inter-process communication for you.
 
 ### Prerequisites
 
-**Client (AV) side:**
-- Python 3.10; preferably managed via [uv](https://docs.astral.sh/uv/) (alternatively, via [Anaconda](https://anaconda.org/), specifically the [`Miniconda3-py310_25.11.1-1` release version on this page](https://repo.anaconda.com/miniconda/))
-- Rust 1.70+ (for QUIC transport)
-- USB webcams (or video sources)
+- [Docker Engine](https://docs.docker.com/engine/install/) 24.0+ with [Docker Compose V2](https://docs.docker.com/compose/install/)
+- [NVIDIA Container Toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html) (for GPU inference)
```
Comment (akrentsel) [inline README.md]: Ok, this is a *bit* confusing in that ideally we show a quickstart option without GPU reqs? Or at least point people to something here? Otherwise it looks like we *need* gpus to try running.

File: README.md [hunk 2]
```diff
@@ -91,56 +91,46 @@ Running on a GPU-equipped cloud instance (e.g., H100):
 
 For detailed architecture, see [docs/ARCHITECTURE.md](docs/ARCHITECTURE.md).
 
-## Quick Start
+## Quick Start (Docker) — Recommended
+
+The recommended way to run TURBO is with Docker. The Docker setup automatically orchestrates all processes — 2 on the server (QUIC server + model servers) and 3 on the client (client orchestrator + web dashboard + QUIC client) — handling startup ordering, ZMQ socket management, and inter-process communication for you.
 
 ### Prerequisites
```
Comment (akrentsel) [inline README.md:100]: Can you add a line somewhere saying approx how much storage is needed for all the data? I ran out of space on my first attempt at setup. (it took up 15G to clone all...)

File: README.md [hunk 3]
```diff
@@ -152,6 +142,130 @@ For detailed architecture, see [docs/ARCHITECTURE.md](docs/ARCHITECTURE.md).
...
+   wget https://storage.googleapis.com/turbo-nines-2026/full-eval.zip
+
+   # Extract to your home directory (creates ~/full-eval/)
+   unzip full-eval.zip -d ~
+   ```
+
+4. **Generate SSL keys for QUIC:**
+   ```bash
+   cd src/quic
+   pip install cryptography   # if not already installed
```
Comment (akrentsel) [inline README.md]: should we mention somnewhere earlier in thge qucikstart that you need to do something with uv? need an env setup no?

Comment (akrentsel) [reply README.md]: `uv sync` should be first step?

File: README.md [hunk 4]
```diff
@@ -152,6 +142,130 @@ For detailed architecture, see [docs/ARCHITECTURE.md](docs/ARCHITECTURE.md).
...
+   ```
+
+   Edit `docker/.env` and update the following values to match your host system:
+
+   | Variable | Description | Default |
+   |---|---|---|
+   | `HOST_UID` | Your host user ID (run `id -u`) | `1000` |
+   | `HOST_GID` | Your host group ID (run `id -g`) | `1000` |
+   | `EXPERIMENT_OUTPUT_DIR` | Absolute path for experiment output | (must set) |
+   | `EFFDET_MODELS_DIR` | Absolute path to model checkpoints (server) | (must set) |
```
Comment (akrentsel) [inline README.md]: can you give the values for this that would result from cloning as we did above? shouldn't this be clear from the quickstart?

File: README.md [hunk 5]
```diff
@@ -152,6 +142,130 @@ For detailed architecture, see [docs/ARCHITECTURE.md](docs/ARCHITECTURE.md).
...
+   | `MODEL_FULL_EVAL_DIR` | Absolute path to evaluation data (client) | (must set) |
```
Comment (akrentsel) [inline README.md]: same here

File: README.md [hunk 6]
```diff
@@ -152,6 +142,130 @@ For detailed architecture, see [docs/ARCHITECTURE.md](docs/ARCHITECTURE.md).
...
+
+   Most other settings (networking, ports, SSL paths) work out of the box for same-host testing. See [docker/README.Docker.md](docker/README.Docker.md) for the full reference.
+
+6. **Create the experiment output directory:**
+   ```bash
+   mkdir -p ~/experiment2-out
```
Comment (akrentsel) [inline README.md:195]: Wait, this is confusing to do this way. Above you say to specify your own path for this, but here you use this folder which is what was in the default example file.

Let's above tell people to make the output be ~/experiment-out, and then here have that be the mkdir

Comment (akrentsel) [reply README.md:195]: (quickstart should require as little thinking as possible...)

File: README.md [hunk 7]
```diff
@@ -152,6 +142,130 @@ For detailed architecture, see [docs/ARCHITECTURE.md](docs/ARCHITECTURE.md).
...
+### Running
+
+All Docker commands should be run from the `docker/` directory:
+```bash
+cd docker
+```
+
+**Run both client and server on the same host:**
+```bash
+docker compose --profile client --profile server up --build
```
Comment (akrentsel) [inline README.md]: This failed.

```
(turbo) akrentsel@crosscheck-train:~/temp/turbo$ docker --version
Docker version 28.2.2, build 28.2.2-0ubuntu1~24.04.1
```

```
akrentsel@crosscheck-train:~/temp/turbo$ source .venv/bin/activate
(turbo) akrentsel@crosscheck-train:~/temp/turbo$ docker compose --profile client --profile server up --build
unknown flag: --profile
```

My docker version is newer than what you specified above. Is there another setup step necessary?

Comment (akrentsel) [reply README.md]: ended up getting docker compose installed, but ti required adding an extension

File: README.md [hunk 8]
```diff
@@ -152,6 +142,130 @@ For detailed architecture, see [docs/ARCHITECTURE.md](docs/ARCHITECTURE.md).
...
+```
```
Comment (akrentsel) [inline README.md:210]: Failed:  => CANCELED [quic_client final 1/3] FROM docker.io/library/alpine:3.18@sha256:de0eb0b3f2a47ba1eb89389859a9bd88b28e82f5826b6969ad604979713c2d4f                                       0.1s
 => => resolve docker.io/library/alpine:3.18@sha256:de0eb0b3f2a47ba1eb89389859a9bd88b28e82f5826b6969ad604979713c2d4f                                                                  0.0s
 => => sha256:802c91d5298192c0f3a08101aeb5f9ade2992e22c9e27fa8b88eab82602550d0 582B / 582B                                                                                            0.0s
 => => sha256:de0eb0b3f2a47ba1eb89389859a9bd88b28e82f5826b6969ad604979713c2d4f 8.08kB / 8.08kB                                                                                        0.0s
 => => sha256:fd032399cd767f310a1d1274e81cab9f0fd8a49b3589eba2c3420228cd45b6a7 1.02kB / 1.02kB                                                                                        0.0s
 => [quic_server internal] load build context                                                                                                                                         0.1s
 => => transferring context: 139.06kB                                                                                                                                                 0.0s
 => CACHED [rust_base rust_build 2/4] WORKDIR /app                                                                                                                                    0.0s
 => CACHED [rust_base rust_build 3/4] RUN apk add --no-cache clang lld musl-dev git                                                                                                   0.0s
 => ERROR [rust_base rust_build 4/4] RUN --mount=type=bind,source=src/quic/quic_client,target=quic_client     --mount=type=bind,source=src/quic/quic_conn,target=quic_conn     --mou  0.0s
------
 > [rust_base rust_build 4/4] RUN --mount=type=bind,source=src/quic/quic_client,target=quic_client     --mount=type=bind,source=src/quic/quic_conn,target=quic_conn     --mount=type=bind,source=src/quic/quic_server,target=quic_server     --mount=type=bind,source=src/quic/Cargo.toml,target=Cargo.toml     --mount=type=bind,source=src/quic/Cargo.lock,target=Cargo.lock     --mount=type=bind,source=./src/quic/ssl_cert.pem,target=ssl_cert.pem     --mount=type=bind,source=./src/quic/ssl_key.pem,target=ssl_key.pem     --mount=type=cache,target=/app/target/     --mount=type=cache,target=/usr/local/cargo/git/db     --mount=type=cache,target=/usr/local/cargo/registry/     cargo build --locked --release &&     IFS=',' &&     for app_name in client,server; do     echo $app_name && echo $(ls ./target/release) && cp ./target/release/$app_name /bin/$app_name;     done:
------
[+] up 0/7
 ⠙ Image docker-quic_server           Building                                                                                                                                         5.4s
 ⠙ Image docker-client_python_main    Building                                                                                                                                         5.4s
 ⠙ Image docker-server_python_main    Building                                                                                                                                         5.4s
 ⠙ Image docker-client_python_monitor Building                                                                                                                                         5.4s
 ⠙ Image docker-rust_base             Building                                                                                                                                         5.4s
 ⠙ Image docker-python_base           Building                                                                                                                                         5.4s
 ⠙ Image docker-quic_client           Building                                                                                                                                         5.4s
Dockerfile_turbo_rust_base:22

--------------------

  21 |     # output directory before the cache mounted /app/target is unmounted.

  22 | >>> RUN --mount=type=bind,source=src/quic/quic_client,target=quic_client \

  23 | >>>     --mount=type=bind,source=src/quic/quic_conn,target=quic_conn \

  24 | >>>     --mount=type=bind,source=src/quic/quic_server,target=quic_server \

  25 | >>>     --mount=type=bind,source=src/quic/Cargo.toml,target=Cargo.toml \

  26 | >>>     --mount=type=bind,source=src/quic/Cargo.lock,target=Cargo.lock \

  27 | >>>     --mount=type=bind,source=${SSL_CERT_PATH},target=ssl_cert.pem \

  28 | >>>     --mount=type=bind,source=${SSL_KEY_PATH},target=ssl_key.pem \

  29 | >>>     --mount=type=cache,target=/app/target/ \

  30 | >>>     --mount=type=cache,target=/usr/local/cargo/git/db \

  31 | >>>     --mount=type=cache,target=/usr/local/cargo/registry/ \

  32 | >>>     cargo build --locked --release && \

  33 | >>>     IFS=',' && \

  34 | >>>     for app_name in $APP_NAME_LIST; do \

  35 | >>>     echo $app_name && echo $(ls ./target/release) && cp ./target/release/$app_name /bin/$app_name; \

  36 | >>>     done

  37 |

--------------------

target quic_server: failed to solve: failed to compute cache key: failed to calculate checksum of ref a276119d-5052-4220-8bec-755d6b28747c::zc7a4c38u00ft7tc921yzbqcv: "/src/quic/Cargo.lock": not found

File: README.md (hunk 7 above)
Comment (hwei0) [reply README.md]: yeah its this plugin https://docs.docker.com/compose/install/linux/; i just added it to README as a requirement!

File: README.md (hunk 8 above)
Comment (hwei0) [reply README.md:210]: resolved by removing `Cargo.lock` from `.gitignore`
